cpp/src/bitmask/bit_mask.h

```diff
@@ -0,0 +1,290 @@
+   *                                    values indicate un-initialized.  Default is uninitialized
+   *  @param[in]  padding_bytes         optional, specifies byte boundary the data should be padded to.
+   *                                    Defaults to 64 bytes, meaning the space allocated will be rounded
+   *                                    up to the next multiple of 64 bytes.
+   *
+   *  @return GDF_SUCCESS on success, the RMM or CUDA error on error
+   */
+  gdf_error create_bit_mask(bit_mask_t **mask, gdf_size_type number_of_records, int fill_value = -1, gdf_size_type padding_bytes = 64) {
+    //
+    //  To handle padding, we will round the number_of_records up to the next padding boundary, then identify how many element
+    //  that equates to.  Then we can allocate the appropriate amount of storage.
+    //
+    gdf_size_type num_bytes = (number_of_records + 7) / 8;
+    gdf_size_type num_padding_blocks = (num_bytes + padding_bytes - 1) / padding_bytes;
+    gdf_size_type num_elements = bit_mask::num_elements(num_padding_blocks * 8 * padding_bytes);
+
+    RMM_TRY(RMM_ALLOC(mask, sizeof(bit_mask_t) * num_elements, 0));
+
+    if (fill_value == 0) {
+      CUDA_TRY(cudaMemset(*mask, 0, sizeof(bit_mask_t) * num_elements));
+    } else if (fill_value == 1) {
+      //
+      //  Filling with one is a more complex case, as we want the first
+      //  number_of_records bits set to 1, and all of the remaining
+      //  bits set to 0.
+      //
+      //  We accomplish this by identifying the number of used element.
+      //  We'll set all elements from 0 to used_elements - 1 to 1, all elements
+      //  from used_elements to num_elements - 1 to 0.  Finally, the last
+      //  used element will be constructed to set to a left mask of ones
+      //  and stored out on the device.
+      //
+      gdf_size_type used_elements = bit_mask::num_elements(number_of_records);
+      CUDA_TRY(cudaMemset(*mask, 0xff, sizeof(bit_mask_t) * used_elements));
+      CUDA_TRY(cudaMemset(&(*mask)[used_elements], 0, sizeof(bit_mask_t) * (num_elements - used_elements)));
+
+      if ((number_of_records % detail::BITS_PER_ELEMENT) != 0) {
+        //
+        //  Need to worry about the case where we set bits past the end
+        //
+        int bits_used_in_last_element = number_of_records - (used_elements - 1) * detail::BITS_PER_ELEMENT;
+        bit_mask_t temp = bit_mask_t{1} << bits_used_in_last_element;
+        temp--;
+
+        return put_element(temp, &(*mask)[used_elements-1]);
+      }
+    }
+
+    return GDF_SUCCESS;
+  }
+
+  /**
+   *  @brief check to see if the specified bit is set to one
+   *
+   *  @param[in]  valid         The bit mask to update
+   *  @param[in]  record_idx    The record index
+   *
+   *  @return which bit within the bit mask
+   */
+  template <typename T>
+  CUDA_HOST_DEVICE_CALLABLE
+  bool is_valid(const bit_mask_t *valid, T record_idx) {
```

Comment:
Is this supposed to replace the is_valid function which was previously defined in a utilitiy class? Does that mean the other one is going away?

Reply:
The idea is that any and all other bitmask manipulation code would go away in favor of this code.